URLs are parsed and produce records

annevk opened the pull request:
No platinum label just yet, but this changes a lot and not all changes
were done manually so review is appreciated.

This should resolve these bugs:

https://www.w3.org/Bugs/Public/show_bug.cgi?id=26400
https://www.w3.org/Bugs/Public/show_bug.cgi?id=26402

There is definitely room for future cleanup:

* Stop using the “resulting URL string” and “resulting URL record”
concepts in most scenarios. Assigning the result to a variable is
clearer.
* Make encoding a more explicit argument (though keep the appropriate
defaulting).
* Perhaps split “parse a URL” into an algorithm that takes an element
and another that takes a base URL.
* Figure out if more places ought to invoke the URL parser directly (in
particular standalone APIs where we want to use UTF-8 as encoding
rather than defaulting.)

domenic commented on source:
```diff
@@ -2698,16 +2698,16 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
-     <li>The <dfn data-noexport="" data-x="concept-url-query">query</dfn> component of a <span>parsed URL</span>
-     <li>The <dfn data-noexport="" data-x="concept-url-fragment">fragment</dfn> component of a <span>parsed URL</span>
+     <li><dfn data-noexport="" data-x-href="https://url.spec.whatwg.org/#concept-url">URL record</dfn>
+     <li>The <dfn data-noexport="" data-x="concept-url-scheme">scheme</dfn> component of a <span>URL record</span>
```
Want to add data-x-hrefs to these while you're here?

domenic commented on source:
```diff
@@ -2698,16 +2698,16 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
-     <li>The <dfn data-noexport="" data-x="concept-url-fragment">fragment</dfn> component of a <span>parsed URL</span>
+     <li><dfn data-noexport="" data-x-href="https://url.spec.whatwg.org/#concept-url">URL record</dfn>
+     <li>The <dfn data-noexport="" data-x="concept-url-scheme">scheme</dfn> component of a <span>URL record</span>
+     <li>The <dfn data-noexport="" data-x="concept-url-scheme-data">scheme data</dfn> component of a <span>URL record</span>
```
Maybe nest these in a `<ul>` underneath the `<li>`, similar to what is done with the states above? Then you don't have to repeat "component of a URL record"

domenic commented on source:
```diff
@@ -6430,26 +6430,23 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
-  <p>Resolving a URL is the process of taking a <span>relative URL</span> and obtaining the
-  <span>absolute URL</span> that it implies.</p>
+  <p>Parsing a URL is the process of taking a <span>URL string</span> and obtaining the <span>URL
+  record</span> that it implies. While this process is defined in the URL Standard the HTML Standard
```
comma after "URL Standard"

annevk replied:
Also lowercased the "S" and added "WHATWG" in front to be consistent.

domenic commented:
Found a few "resolves" still left:

- An absolute URL used by APIs called by scripts that use this environment settings object to resolve relative URLs.
- The assign() method uses the entry settings object to resolve the URL
- Since we already know the url string is an absolute URL, parsing it will never return an error, and it doesn't matter what it is resolved relative to.

domenic commented on source:
```diff
@@ -79791,18 +79775,14 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
-
-    <p>If this is not successful, throw a <code>SyntaxError</code> exception and abort these
-    steps.</p>
+   <li><p>Let <var>urlRecord</var> be the result of <span data-x="parse a url">parsing</span>
```
This is not correct; there is no single result of parsing a URL. You need to keep something similar about the resulting URL record.

domenic commented on source:
```diff
@@ -80204,22 +80184,22 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
-       <li><p>If <var>parsed URL</var>'s <span data-x="concept-url-query">query</span>
-       component is not null, then first append a U+003F QUESTION MARK character (?) to <var>script source</var>, and then append <var>parsed URL</var>'s <span
+       <li><p>If <var>URL record</var>'s <span data-x="concept-url-query">query</span>
+       component is not null, then first append a U+003F QUESTION MARK character (?) to <var>script source</var>, and then append <var>URL record</var>'s <span
```
Fix wrapping while you're here

domenic commented on source:
```diff
@@ -80204,22 +80184,22 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+       <li><p>If <var>URL record</var>'s <span
        data-x="concept-url-fragment">fragment</span> component is not null, then first append a
-       U+0023 NUMBER SIGN character (#) to <var>script source</var>, and then append <var>parsed URL</var>'s <span data-x="concept-url-fragment">fragment</span> component to
+       U+0023 NUMBER SIGN character (#) to <var>script source</var>, and then append <var>URL record</var>'s <span data-x="concept-url-fragment">fragment</span> component to
```
Fix wrapping while you're here

domenic commented on source:
```diff
@@ -8673,14 +8660,16 @@ partial /*sealed*/ interface <dfn>Document</dfn> {
-   <span>API base URL</span> specified by the <span>entry settings object</span>. If this is not
-   successful, throw a <code>SyntaxError</code> exception and abort these steps. Otherwise, let
-   <var>url</var> be the <span>resulting parsed URL</span>.</p></li>
+   <li><p>Let <var>urlRecord</var> be the result of <span data-x="parse a url">parsing</span>
```
The previous structure was clearer. This is not correct, since the algorithm does not produce a url record; you have to explicitly reference resulting URL record.

annevk commented:
Addressed these nits. If I have time over later today I can look into the grander rewrites we discussed on IRC.

I also removed scheme data by the way, though we should file a follow up on javascript URL parsing. It basically has to be a special case I think.

annevk commented:
#532 is the "scheme data" issue that is now fixed by this change.

annevk commented:
I did not get to the grander rewrite, but I'm still interested in investigating changing the second argument of parse a URL to document/environment settings object. Provided you still think it would be clearer to write out element's node document everywhere.

annevk commented:
Ah, you are correct. perhaps I should look into if we can avoid returning two values, since that seems rather poor practice.

domenic commented:
> Provided you still think it would be clearer to write out element's node document everywhere.

Yeah, I still think that would be nice.

> perhaps I should look into if we can avoid returning two values, since that seems rather poor practice.

Yeah, just returning the record in most cases seems better, and then explicitly serializing by calling into the URL Standard when appropriate. But I'm not sure how much work that is compared to just going back to the old phrasing (with slightly updated terminology).

annevk commented:
Not returning two values was too hard at this point. We should slowly move to use URL records everywhere and serialize them only when needed (e.g., the `<img>` section has several violations of this) and then make that change.

I also think Worker and EventSource should be updated to use the URL parser directly. The way EventSource claims its second argument is optional is also bogus (since a dictionary will always be passed thanks to IDL). I plan to do some follow up for these once this is landed (easier to refactor then).

annevk commented:
I merged "fixup" into the first commit btw and created "fixup2" for all changes since then. Should probably heft left "fixup" alone, but so be it.

domenic commented:
Lovely! LGTM with a couple comments:

- commit message needs to be updated to reflect the broader scope of the PR and the cleanup that has taken place so far and the scheme data issue being fixed
- I don't understand what "no platinum label yet" means
- Although the spec has been changing a bit, in general I think we prefer multi-word variable names like _URL string_ and _URL record_ instead of _urlString_ and _urlRecord_. But If you want to move toward the latter that's fine too.

annevk commented:
I would prefer moving towards single-word variables. I've often found the multi-word ones confusing to read as it's unclear what's part of the variable and what is part of the sentence.
